### PR TITLE
Support non-float global phase in QPY

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit adds support for non-float global phase types in qpy.
Previously, the QPY format only supported global phase parameters as
floats, however the QuantumCircuit object supports additional types,
like int, parameter expressions, etc. To support this in QPY a format
revision is needed as this was not expressible in QPYv1. This commit
adds a new format QPY v2 which makes this change to the circuit HEADER
struct so we can represent different types for global phase. This will
be released with 0.19.0, so any additional format changes we make in
0.19.0 will be also part of QPY v2.

### Details and comments

TODO:

- [x] Add release note
- [x] Add backwards compat tests for qiskit >=0.19.0 with new global phase types

### CODE REVIEW

File: qiskit/circuit/qpy_serialization.py [hunk 1]
```diff
@@ -367,6 +411,22 @@ class if it's defined in Qiskit. Otherwise it falls back to the custom
...
+        "global_phase_type",
+        "global_phase_size",
+        "num_qubits",
+        "num_clbits",
+        "metadata_size",
+        "num_registers",
+        "num_instructions",
+    ],
+)
+HEADER_V2_PACK = "!H1cHIIQIQ"
```
Comment (jakelishman) [inline qiskit/circuit/qpy_serialization.py:428]: Does the explicit `1c` in the string have different semantics to just `c`?

File: qiskit/circuit/qpy_serialization.py [hunk 2]
```diff
@@ -85,6 +87,49 @@
...
+------
+
+The contents of HEADER are defined as a C struct are:
+
+.. code-block:: c
+
+    struct {
+        unit16_t name_size;
+        char global_phase_type;
+        unit16_t global_phase_size;
```
Comment (jakelishman) [inline qiskit/circuit/qpy_serialization.py:109]:
```suggestion
        uint16_t name_size;
        char global_phase_type;
        uint16_t global_phase_size;
```

File: qiskit/circuit/qpy_serialization.py [hunk 3]
```diff
@@ -85,6 +87,49 @@
...
+    }
+
+This is immediately followed by ``name_size`` bytes of utf8 data for the name
+of the circuit. Following this is immediately ``global_phase_size`` bytes
+representing the global phase. The content of that data is dictated by the
+value of ``global_phase_type``. If it's ``'f'`` the data is a float and is the
+size of a ``double``. If it's ``'p'`` defines a :class:`~qiskit.circuit.Parameter`
+object  which is represented by a PARAM struct (see below), ``e`` defines a
+:class:`~qiskit.circuit.ParameterExpression` object (that's not a
+:class:`~qiskit.circuit.Paramter`) which is represented by a PARAM_EXPR struct
```
Comment (jakelishman) [inline qiskit/circuit/qpy_serialization.py:125]:
```suggestion
:class:`~qiskit.circuit.Parameter`) which is represented by a PARAM_EXPR struct
```

File: releasenotes/notes/qpy-v2-f1c380b40936cccf.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+---
+features:
+  - |
+    QPY serialization is now capable of representing
+    :attr:`~qiskit.circuit.QuantumCircuit.global_phase` attributes of a
+    :class:`~qiskit.circuit.QuantumCircuit` object that are a ``int``,
+    :class:`~qiskit.circuit.Parameter` object, or
+    :class:`~qiskit.circuit.ParameterExpression object`. Previous versions of
```
Comment (jakelishman) [inline releasenotes/notes/qpy-v2-f1c380b40936cccf.yaml:8]:
```suggestion
    :class:`~qiskit.circuit.ParameterExpression` object. Previous versions of
```

**Review by jakelishman — COMMENTED** [review]
Looks sensible overall.  The general principle looks solid.  I saw a couple of typos in the documentation, and one minor oddness (perhaps?) with a struct packing string, but that's it.

File: qiskit/circuit/qpy_serialization.py (hunk 1 above)
Comment (mtreinish) [reply qiskit/circuit/qpy_serialization.py:428]: No they should be the same, `1c` means just 1 repeat of `c`. It was just a habit from when I was using fixed length strings in the first draft of QPY. It makes more sense if you do something like `4c` which is the same as `cccc`. It's documented here: https://docs.python.org/3/library/struct.html#format-characters (it's the first paragraph after the numbered `Notes` list). I can change it to just be `c` if you'd prefer that.

Comment (jakelishman) [reply qiskit/circuit/qpy_serialization.py:428]: I have no preference, I was just wondering if a count of one might have implied something else.  The idea of fixed-length C strings seems totally valid to me, and I'm happy the way it is.

**Review by jakelishman — DISMISSED** [review]

File: releasenotes/notes/qpy-v2-f1c380b40936cccf.yaml [hunk 5]
```diff
@@ -0,0 +1,13 @@
+---
+features:
+  - |
+    QPY serialization is now capable of representing
+    :attr:`~qiskit.circuit.QuantumCircuit.global_phase` attributes of a
+    :class:`~qiskit.circuit.QuantumCircuit` object that are a ``int``,
+    :class:`~qiskit.circuit.Parameter` object, or
+    :class:`~qiskit.circuit.ParameterExpression` object. Previous versions of
+    QPY would only accept a :attr:`~qiskit.circuit.QuantumCircuit.global_phase`
+    that were ``float``s.
```
Comment (jakelishman) [inline releasenotes/notes/qpy-v2-f1c380b40936cccf.yaml:10]:
```suggestion
    that was a ``float``.
```

I _think_ Sphinx doesn't recognise ` `` ` followed by a word character, but not 100% sure.

Comment (mtreinish) [reply releasenotes/notes/qpy-v2-f1c380b40936cccf.yaml:10]: Yeah, sphinx doesn't like that. Good catch.

**Review by jakelishman — APPROVED** [review]

**Review by jakelishman — APPROVED** [review]

**Review by Cryoris — APPROVED** [review]
It fixes the bug we encountered with QAOA / the EvolvedOperator ansatz (also without the fix we added for that 😄).

**Pull request merged**